core/node/metadata_calculator/src/recovery/mod.rs

```diff
@@ -279,13 +313,18 @@ impl AsyncTreeRecovery {
 
         let finalize_latency = RECOVERY_METRICS.latency[&RecoveryStage::Finalize].start();
         let actual_root_hash = tree.root_hash().await;
-        anyhow::ensure!(
-            actual_root_hash == snapshot.expected_root_hash,
-            "Root hash of recovered tree {actual_root_hash:?} differs from expected root hash {:?}. \
-             If pruning is enabled and the tree is initialized some time after node recovery, \
-             this is caused by snapshot storage logs getting pruned; this setup is currently not supported",
-            snapshot.expected_root_hash
-        );
+        if let Some(expected_root_hash) = init_params.expected_root_hash {
```

Comment:
It would be beneficial to record the state root hash of the last hard-pruned L1 batch somewhere (in pruning logs?) so that it can be checked here. If this sounds OK, I'd prefer to implement it in a follow-up PR.